rts/motoko-rts/src/idl.rs

```diff
@@ -486,3 +512,310 @@ unsafe extern "C" fn skip_fields(tb: *mut Buf, buf: *mut Buf, typtbl: *mut *mut
         *n -= 1;
     }
 }
+
+unsafe fn is_null_opt_reserved(typtbl: *mut *mut u8, end: *mut u8, t: i32) -> bool {
+    if is_primitive_type(t) {
+        return t == IDL_PRIM_null || t == IDL_PRIM_reserved;
+    }
+
+    // unfold t
+    let mut t = t;
+
+    let mut tb = Buf {
+        ptr: *typtbl.add(t as usize),
+        end: end,
+    };
+
+    t = sleb128_decode(&mut tb);
+
+    return t == IDL_CON_opt;
+}
+
+// https://github.com/dfinity/candid/blob/master/rust/candid/src/types/subtype.rs#L10
+// https://github.com/dfinity/candid/blob/20b84d1c1515e2c1db353ebe02b738486f835466/spec/Candid.md
+unsafe fn sub(
+    rel: &BitRel,
+    p: bool,
+    typtbl1: *mut *mut u8,
+    typtbl2: *mut *mut u8,
+    end1: *mut u8,
+    end2: *mut u8,
+    t1: i32,
+    t2: i32,
+) -> bool {
+    if t1 >= 0 && t2 >= 0 {
+        let t1 = t1 as u32;
+        let t2 = t2 as u32;
+        if rel.get(p, t1, t2) {
+            // cached: succeed!
+            return true;
+        };
+        // cache and continue
+        rel.set(p, t1, t2);
+    };
+
+    // re-declare as mut for any unfolding
+    let mut t1 = t1;
+    let mut t2 = t2;
+
+    /* primitives reflexive */
+    if is_primitive_type(t1) && is_primitive_type(t2) && t1 == t2 {
+        return true;
+    }
+
+    // unfold t1, if necessary
+    let mut tb1 = Buf {
+        ptr: if t1 < 0 {
+            end1
+        } else {
+            *typtbl1.add(t1 as usize)
+        },
+        end: end1,
+    };
+
+    if t1 >= 0 {
+        t1 = sleb128_decode(&mut tb1);
+    };
+
+    // unfold t2, if necessary
+    let mut tb2 = Buf {
+        ptr: if t2 < 0 {
+            end2
+        } else {
+            *typtbl2.add(t2 as usize)
+        },
+        end: end2,
+    };
+
+    if t2 >= 0 {
+        t2 = sleb128_decode(&mut tb2);
+    };
+
+    match (t1, t2) {
+        (_, IDL_CON_alias) | (IDL_CON_alias, _) => idl_trap_with("sub: unexpected alias"),
```

Comment:
It's used by extended candid only and encodes mutable data. The subtyping test is currently only implemented for, and used on, pure candid types, never extended ones. So this could really be an assert.